### ISSUE TITLE
Rule Change: improve no-misleading-character-class so that it reports on the sub-node range rather than the entire node

### ISSUE THREAD

**Issue by bradzacher** [eslint/eslint#16682]
### What rule do you want to change?

no-misleading-character-class

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
/[\u{2ea}-\u{2eb}\u{1100}-\u{11ff}\u{2e80}-\u{2e99}\u{2e9b}-\u{2ef3}\u{2f00}-\u{2fd5}\u{3005}\u{3007}\u{3021}-\u{3029}\u{302e}-\u{302f}\u{3038}-\u{303a}\u{303b}\u{3041}-\u{3096}\u{309d}-\u{309e}\u{309f}\u{30a1}-\u{30fa}\u{30fd}-\u{30fe}\u{30ff}\u{3105}-\u{312f}\u{3131}-\u{318e}\u{31a0}-\u{31bf}\u{31f0}-\u{31ff}\u{3200}-\u{321e}\u{3260}-\u{327e}\u{32d0}-\u{32fe}\u{3300}-\u{3357}\u{3400}-\u{4dbf}\u{4e00}-\u{9ffc}\u{a960}-\u{a97c}\u{ac00}-\u{d7a3}\u{d7b0}-\u{d7c6}\u{d7cb}-\u{d7fb}\u{f900}-\u{fa6d}\u{fa70}-\u{fad9}\u{ff66}-\u{ff6f}\u{ff71}-\u{ff9d}\u{ffa0}-\u{ffbe}\u{ffc2}-\u{ffc7}\u{ffca}-\u{ffcf}\u{ffd2}-\u{ffd7}\u{ffda}-\u{ffdc}\u{16ff0}-\u{16ff1}\u{1b000}\u{1b001}-\u{1b11e}\u{1b150}-\u{1b152}\u{1b164}-\u{1b167}\u{1f200}\u{20000}-\u{2a6dd}\u{2a700}-\u{2b734}\u{2b740}-\u{2b81d}\u{2b820}-\u{2cea1}\u{2ceb0}-\u{2ebe0}\u{2f800}-\u{2fa1d}\u{30000}-\u{3134a}]/u;
```


### What does the rule currently do for this code?

The rule currently errors on the entire RegExp literal.
[eslint playground example](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzOiBbXCJlcnJvclwiXSAqL1xuY29uc3QgcmVnZXggPSAvW1xcdXsyZWF9LVxcdXsyZWJ9XFx1ezExMDB9LVxcdXsxMWZmfVxcdXsyZTgwfS1cXHV7MmU5OX1cXHV7MmU5Yn0tXFx1ezJlZjN9XFx1ezJmMDB9LVxcdXsyZmQ1fVxcdXszMDA1fVxcdXszMDA3fVxcdXszMDIxfS1cXHV7MzAyOX1cXHV7MzAyZX0tXFx1ezMwMmZ9XFx1ezMwMzh9LVxcdXszMDNhfVxcdXszMDNifVxcdXszMDQxfS1cXHV7MzA5Nn1cXHV7MzA5ZH0tXFx1ezMwOWV9XFx1ezMwOWZ9XFx1ezMwYTF9LVxcdXszMGZhfVxcdXszMGZkfS1cXHV7MzBmZX1cXHV7MzBmZn1cXHV7MzEwNX0tXFx1ezMxMmZ9XFx1ezMxMzF9LVxcdXszMThlfVxcdXszMWEwfS1cXHV7MzFiZn1cXHV7MzFmMH0tXFx1ezMxZmZ9XFx1ezMyMDB9LVxcdXszMjFlfVxcdXszMjYwfS1cXHV7MzI3ZX1cXHV7MzJkMH0tXFx1ezMyZmV9XFx1ezMzMDB9LVxcdXszMzU3fVxcdXszNDAwfS1cXHV7NGRiZn1cXHV7NGUwMH0tXFx1ezlmZmN9XFx1e2E5NjB9LVxcdXthOTdjfVxcdXthYzAwfS1cXHV7ZDdhM31cXHV7ZDdiMH0tXFx1e2Q3YzZ9XFx1e2Q3Y2J9LVxcdXtkN2ZifVxcdXtmOTAwfS1cXHV7ZmE2ZH1cXHV7ZmE3MH0tXFx1e2ZhZDl9XFx1e2ZmNjZ9LVxcdXtmZjZmfVxcdXtmZjcxfS1cXHV7ZmY5ZH1cXHV7ZmZhMH0tXFx1e2ZmYmV9XFx1e2ZmYzJ9LVxcdXtmZmM3fVxcdXtmZmNhfS1cXHV7ZmZjZn1cXHV7ZmZkMn0tXFx1e2ZmZDd9XFx1e2ZmZGF9LVxcdXtmZmRjfVxcdXsxNmZmMH0tXFx1ezE2ZmYxfVxcdXsxYjAwMH1cXHV7MWIwMDF9LVxcdXsxYjExZX1cXHV7MWIxNTB9LVxcdXsxYjE1Mn1cXHV7MWIxNjR9LVxcdXsxYjE2N31cXHV7MWYyMDB9XFx1ezIwMDAwfS1cXHV7MmE2ZGR9XFx1ezJhNzAwfS1cXHV7MmI3MzR9XFx1ezJiNzQwfS1cXHV7MmI4MWR9XFx1ezJiODIwfS1cXHV7MmNlYTF9XFx1ezJjZWIwfS1cXHV7MmViZTB9XFx1ezJmODAwfS1cXHV7MmZhMWR9XFx1ezMwMDAwfS1cXHV7MzEzNGF9XS91O1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6eyJlczYiOnRydWV9fX0=)

### What will the rule do after it's changed?

The rule will report on the sub-node range that violates the rule instead of the entire literal node. For long regexes it can be pretty-well impossible to understand what the actual problem is if the rule reports on the entire node.

Example with reporting on the relevant violating sub-node ranges:
```js
/[\u{2ea}-\u{2eb}\u{1100}-\u{11ff}\u{2e80}-\u{2e99}\u{2e9b}-\u{2ef3}\u{2f00}-\u{2fd5}\u{3005}\u{3007}\u{3021}-\u{3029}\u{302e}-\u{302f}\u{3038}-\u{303a}\u{303b}\u{3041}-\u{3096}\u{309d}-\u{309e}\u{309f}\u{30a1}-\u{30fa}\u{30fd}-\u{30fe}\u{30ff}\u{3105}-\u{312f}\u{3131}-\u{318e}\u{31a0}-\u{31bf}\u{31f0}-\u{31ff}\u{3200}-\u{321e}\u{3260}-\u{327e}\u{32d0}-\u{32fe}\u{3300}-\u{3357}\u{3400}-\u{4dbf}\u{4e00}-\u{9ffc}\u{a960}-\u{a97c}\u{ac00}-\u{d7a3}\u{d7b0}-\u{d7c6}\u{d7cb}-\u{d7fb}\u{f900}-\u{fa6d}\u{fa70}-\u{fad9}\u{ff66}-\u{ff6f}\u{ff71}-\u{ff9d}\u{ffa0}-\u{ffbe}\u{ffc2}-\u{ffc7}\u{ffca}-\u{ffcf}\u{ffd2}-\u{ffd7}\u{ffda}-\u{ffdc}\u{16ff0}-\u{16ff1}\u{1b000}\u{1b001}-\u{1b11e}\u{1b150}-\u{1b152}\u{1b164}-\u{1b167}\u{1f200}\u{20000}-\u{2a6dd}\u{2a700}-\u{2b734}\u{2b740}-\u{2b81d}\u{2b820}-\u{2cea1}\u{2ceb0}-\u{2ebe0}\u{2f800}-\u{2fa1d}\u{30000}-\u{3134a}]/u;
//                                                                                                            ^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ^^^^^^^^^^^^^^^^^^
```

This will result in more lint rule reports because it will report more than once per node.
But I think the increased understandability of the reports is worth it.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

_No response_

**Comment by mdjermanovic** [comment]
Make sense to me because the current reporting from this rule really isn't informative enough. I'll work on this.
